OpenRA.Mods.Common/Widgets/Logic/Lobby/MapPreviewLogic.cs

```diff
@@ -44,220 +51,258 @@ public class MapPreviewLogic : ChromeLogic
+						modData.Translation.GetString(Downloading, Translation.Arguments("size", map.DownloadBytes / 1024));
 
-					retry.OnClick = () =>
-					{
-						modData.MapCache.UpdateMaps();
-						var (map, _) = getMap();
-						if (map.Status == MapStatus.DownloadError)
-						{
-							map.Install(mapRepository, () =>
-							{
-								if (orderManager != null)
-									Game.RunAfterTick(() => orderManager.IssueOrder(Order.Command($"state {Session.ClientState.NotReady}")));
-							});
-						}
-						else if (map.Status == MapStatus.Unavailable)
-							modData.MapCache.QueryRemoteMapDetails(mapRepository, new[] { map.Uid });
-					};
-
-					retry.GetText = () => getMap().Map.Status == MapStatus.DownloadError
-						? modData.Translation.GetString(RetryInstall)
-						: modData.Translation.GetString(RetrySearch);
-				}
+				return modData.Translation.GetString(DownloadingPercentage, Translation.Arguments("size", map.DownloadBytes / 1024, "progress", map.DownloadPercentage));
+			};
+
+			previewWidgets[PreviewStatus.Downloading] = new Widget[] { previewSmall, downloadingContainer };
+
+			retryButton = widget.Get<ButtonWidget>("MAP_RETRY");
+			retryButton.OnClick = () =>
+			{
+				var (map, _) = getMap();
 
-				var progressbar = progress.GetOrNull<ProgressBarWidget>("MAP_PROGRESSBAR");
-				if (progressbar != null)
+				var uid = modData.MapCache.GetUpdatedMap(map.Uid);
+				mapUpdateAvailable = mapUpdatesEnabled && uid != null && map.Uid != uid;
+
+				if (map.Status == MapStatus.DownloadError)
 				{
-					progressbar.IsIndeterminate = () => getMap().Map.DownloadPercentage == 0;
-					progressbar.GetPercentage = () => getMap().Map.DownloadPercentage;
-					progressbar.IsVisible = () => getMap().Map.Status == MapStatus.Downloading;
+					map.Install(mapRepository, () =>
+					{
+						if (orderManager != null)
+							Game.RunAfterTick(() => orderManager.IssueOrder(Order.Command($"state {Session.ClientState.NotReady}")));
+					});
 				}
-			}
+				else if (map.Status == MapStatus.Unavailable)
+					modData.MapCache.QueryRemoteMapDetails(mapRepository, new[] { map.Uid });
+			};
+
+			var retryInstall = modData.Translation.GetString(RetryInstall);
+			var retrySearch = modData.Translation.GetString(RetrySearch);
+			retryButton.GetText = () => getMap().Map.Status == MapStatus.DownloadError ? retryInstall : retrySearch;
+
+			previewWidgets[PreviewStatus.Unavailable] = new Widget[] { previewSmall, widget.Get("MAP_STATUS_UNAVAILABLE"), retryButton };
+			previewWidgets[PreviewStatus.DownloadError] = new Widget[] { previewSmall, widget.Get("MAP_STATUS_ERROR"), retryButton };
+			previewWidgets[PreviewStatus.UpdateAvailable] = new Widget[] { previewSmall, widget.Get("MAP_UPDATE_FOUND"), updateButton };
+
+			// Hide all widgets
+			foreach (var preview in previewWidgets)
+				foreach (var p in preview.Value)
+					p.IsVisible = () => false;
+		}
+
+		Widget[] visibleWidgets = Array.Empty<Widget>();
+
+		void UpdateStatus()
```

Comment:
This should be called `UpdateVisibility`, the status variable is updated in `Tick`.